gadopt/energy_solver.py

```diff
@@ -53,82 +50,317 @@
+      solver_parameters:
+        Dicitionary of solver parameters or a string specifying a default configuration
+        provided to PETSc
 
     """
 
+    terms_mapping = {
+        "advection": scalar_eq.advection_term,
+        "diffusion": scalar_eq.diffusion_term,
+        "sink": scalar_eq.sink_term,
+        "source": scalar_eq.source_term,
+    }
+
     def __init__(
         self,
-        T: Function,
-        u: Function,
-        approximation: BaseApproximation,
-        delta_t: Constant,
+        scalar_field: fd.Function,
+        delta_t: fd.Constant,
         timestepper: RungeKuttaTimeIntegrator,
-        bcs: Optional[dict[int, dict[str, Number]]] = None,
-        solver_parameters: Optional[dict[str, Any]] = None,
-        su_advection: bool = False,
-    ):
-        self.T = T
-        self.Q = T.function_space()
-        self.mesh = self.Q.mesh()
+        *,
+        solution_old: Optional[fd.Function] = None,
+        bcs: dict[int, dict[str, Number]] = {},
+        solver_parameters: Optional[dict[str, str | Number] | str] = None,
+    ) -> None:
+        self.solution = scalar_field
         self.delta_t = delta_t
+        self.timestepper = timestepper
 
-        if solver_parameters is None:
-            if self.mesh.topological_dimension() == 2:
-                self.solver_parameters = direct_energy_solver_parameters.copy()
-                if INFO >= log_level:
-                    # not really "informative", but at least we get a 1-line message saying we've passed the energy solve
-                    self.solver_parameters['ksp_converged_reason'] = None
-            else:
-                self.solver_parameters = iterative_energy_solver_parameters.copy()
-                if DEBUG >= log_level:
-                    self.solver_parameters['ksp_monitor'] = None
-                elif INFO >= log_level:
-                    self.solver_parameters['ksp_converged_reason'] = None
-        else:
+        self.Q = scalar_field.function_space()
+        self.mesh = self.Q.mesh()
+        self.solution_old = solution_old or fd.Function(self.Q)
+
+        self.set_boundary_conditions(bcs)
+
+        if isinstance(solver_parameters, dict):
             self.solver_parameters = solver_parameters
-        apply_strongly = is_continuous(T)
+        else:
+            self.set_solver_parameters(solver_parameters)
+
+        # Solver object is set up later to permit editing default solver parameters.
+        self._solver_setup = False
+
+    def set_boundary_conditions(self, bcs: dict[int, dict[str, Number]]) -> None:
+        """Sets up boundary conditions."""
         self.strong_bcs = []
         self.weak_bcs = {}
-        bcs = bcs or {}
-        for id, bc in bcs.items():
+
+        apply_strongly = is_continuous(self.solution)
+
+        for bc_id, bc in bcs.items():
             weak_bc = {}
             for type, value in bc.items():
-                if type == 'T':
+                if type == "T":
```

Comment:
Just a small suggestion but probably strong dirichlet bcs in the generic transport solver should be something like q_strong? or Q? (but that is sometimes used for function spaces...) 

And then energy solver can accept T but it gets passed to generic transport solver...

Reply:
I think this is possible through a class variable, e.g. `self.strong_bcs_tag = "T"`. Then, you would have `if bc_type == self.strong_bcs_tag:`. Is that desirable though?

Reply:
@w3168 Any updates on this?

Reply:
I think self.strong_bcs_tag is ok. you could simplify it to `self.strong_bc`